Quaver.Shared/Screens/Select/SelectScreen.cs

```diff
@@ -542,5 +573,184 @@ public void SelectRandomMap()
+
+                    // Run deletion & Reload in the background.
+                    ThreadScheduler.Run(() => DeleteAndReloadMaps(filePath));
+
+                    // Finally show confirmation notification
+                    NotificationManager.Show(NotificationLevel.Success, $"Successfully deleted {selectedMapset.Title} from Quaver!");
+                }
+                catch (ArgumentOutOfRangeException)
+                {
+                    // Handle error
+                    NotificationManager.Show(NotificationLevel.Error, "An error has occured. Please check runtime.log!");
+                }
+            });
+
+            // Finally show the confirmation dialog that orchestrates the deleting process.
+            DialogManager.Show(confirmDelete);
+        }
+
+        /// <summary>
+        ///     Used to delete a mapset or difficulty, then reload the available maps.
+        /// </summary>
+        /// <param name="path"></param>
+        private void DeleteAndReloadMaps(string path)
+        {
+            var view = View as SelectScreenView;
+            var type = view.ActiveContainer;
+
+            if (view.MapsetScrollContainer.SelectedMapsetIndex < 0)
+                return;
+
+            // Externally loaded map check
+            if (MapManager.Selected.Value.Game != MapGame.Quaver)
+            {
+                // Display error message
+                NotificationManager.Show(NotificationLevel.Error,"This map was loaded from an external source, it cannot be deleted.");
+                return;
+            }
+
+            var preservedMapsetIndex = view.MapsetScrollContainer.SelectedMapsetIndex;
+            var preservedDifficultyIndex = view.DifficultyScrollContainer.SelectedMapIndex;
+
+            try
+            {
+                DeletePath(path);
+
+                // Remove mapset/difficulty from cache.
+                switch (type)
+                {
+                    case SelectContainerStatus.Mapsets:
+                        AvailableMapsets[view.MapsetScrollContainer.SelectedMapsetIndex].Maps.ForEach(MapDatabaseCache.RemoveMap);
+                        break;
+
+                    case SelectContainerStatus.Difficulty:
+                        MapDatabaseCache.RemoveMap(AvailableMapsets[view.MapsetScrollContainer.SelectedMapsetIndex].Maps[view.DifficultyScrollContainer.SelectedMapIndex]);
+                        break;
+
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+
+                MapDatabaseCache.OrderAndSetMapsets();
+
+                if (PreviousSearchTerm.Length > 0)
+                {
+                    // Grab the mapsets available to the user according to their previous search term.
+                    AvailableMapsets = MapsetHelper.SearchMapsets(MapManager.Mapsets, PreviousSearchTerm);
+                    AvailableMapsets = MapsetHelper.OrderMapsetsByConfigValue(AvailableMapsets);
```

Comment:
When you remove a mapset the ordering or the search does not change, so it is unnecessary to do the whole thing all over. Besides, isn't this racing with literally everything in the main thread? I suspect that simplifying this to a simple removal and moving to the first function will also provide a proper fix for spamming delete-enter which doesn't drop these inputs because "mapset deletion is in progress".